carnot/carnot_vote_aggregation.py

```diff
@@ -0,0 +1,421 @@
+        """
+        pass
+
+
+class Carnot2(Carnot):
+    def __init__(self, _id: Id, overlay=Overlay2()):
+        self.latest_committed_block = None
+        self.id: Id = _id
+        self.current_view: View = 0
+        self.highest_voted_view: View = -1
+        self.local_high_qc: Type[Qc] = None
+        self.safe_blocks: Dict[Id, Block] = dict()
+        self.last_view_timeout_qc: Type[AggregateQc] = None
+        self.overlay: Overlay = overlay
+
+    @abstractmethod
+    def commit_block(self, block: Block) -> bool:
+
+        pass
+
+    # Commit the grandparent and all its uncommitted ancestors
+    def commit_the_chain(self, grand_parent: Block):
+        # Create an empty stack to store the blocks in reverse order
+        block_stack = []
+
+        # Start with the grand_parent block
+        current_block = grand_parent
+
+        # Push blocks onto the stack until we reach the parent of the latest_committed_block
+        while current_block != self.latest_committed_block.parent():
+            block_stack.append(current_block)
+            current_block = self.safe_blocks.get(current_block.parent())
+        # Pop and commit blocks from the stack to execute them in order
+        while block_stack:
+            block_to_commit = block_stack.pop()
+            # Commit the transactions of the block using your commit_block method
+            self.commit_block(block_to_commit)
+        latest_committed_block = block_to_commit
+        # Update the latest committed block to be the latest_committed_block
+        self.latest_committed_block = latest_committed_block
+
+    # The  check for the first block generated after unhappy path is added.
+    def block_is_safe(self, block: Block) -> bool:
+        if isinstance(block.qc, StandardQc):
+            return block.view_num == block.qc.view() + 1
+        elif isinstance(block.qc, AggregateQc):
+            return block.view_num == block.qc.view() + 1 and block.extends(self.latest_committed_block())
+        else:
+            return False
+
+    # Update the view for any QC with higher view than the current view.
+    def update_high_qc(self, qc: Qc):
+        match (self.local_high_qc, qc):
+            case (None, new_qc) if isinstance(new_qc, StandardQc):
+                # Set local high QC to the new StandardQc
+                self.local_high_qc = new_qc
+            case (None, new_qc) if isinstance(new_qc, AggregateQc):
+                # Set local high QC to the high QC from the new AggregateQc
+                self.local_high_qc = new_qc.high_qc()
+            case (old_qc, new_qc) if isinstance(new_qc, StandardQc) and new_qc.view > old_qc.view:
+                # Update local high QC if the new StandardQc has a higher view
+                self.local_high_qc = new_qc
+            case (old_qc, new_qc) if isinstance(new_qc,
+                                                AggregateQc) and new_qc.high_qc().view != old_qc.view and new_qc.view > old_qc.view:
```

Comment:
`new_qc.high_qc().view != old_qc.view` is redundant and implied by `new_qc.view > old_qc.view`.
This is different from standard Carnot, could you explain why?